src/main/java/org/hydev/mcpm/client/injector/PluginLoader.java

```diff
@@ -1,45 +1,135 @@
     {
-        // TODO: Implement this
-        throw new UnsupportedOperationException("TODO");
+        // 2. Load plugin
+        var pm = Bukkit.getPluginManager();
+        try
+        {
+            var plugin = pm.loadPlugin(jar);
+            assert plugin != null;
+
+            // 3. Call onLoad()
+            plugin.onLoad();
+            pm.enablePlugin(plugin);
+        }
+        catch (InvalidPluginException | InvalidDescriptionException e)
+        {
+            // These are errors indicating that the plugin we're trying to load is badly formatted.
+            // There are nothing we can do.
+            e.printStackTrace();
+        }
     }
 
-    /**
-     * Dynamically unload a local plugin through JVM reflections and classloader hacks
-     *
-     * @param name Loaded plugin name
-     */
-    public void unloadPlugin(String name)
+    @Override
+    public void unloadPlugin(String name) throws PluginNotFoundException
     {
-        // TODO: Implement this
-        throw new UnsupportedOperationException("TODO");
+        var pm = Bukkit.getPluginManager();
+
+        // 1. Find plugin by name
+        var plugin = Arrays.stream(pm.getPlugins()).filter(p -> p.getName().equalsIgnoreCase(name)).findFirst()
+            .orElseThrow(() -> new PluginNotFoundException(name));
+
+        // 2. Unload plugin
+        pm.disablePlugin(plugin);
+
+        // 3. Unregister listeners
+        getPrivateField(pm, "listeners", new TypeToken<Map<Event, SortedSet<RegisteredListener>>>(){})
+            .ifPresent(listeners -> listeners.values().forEach(set -> set.removeIf(l -> l.getPlugin() == plugin)));
+
+        // 4. Remove command from command map
+        getPrivateField(pm, "commandMap", new TypeToken<SimpleCommandMap>(){}).ifPresent(cmdMap ->
+            getPrivateField(pm, "knownCommands", new TypeToken<Map<String, Command>>(){}).ifPresent(cmds ->
+                cmds.values().stream()
+                    .filter(cmd -> cmd instanceof PluginCommand).map(cmd -> (PluginCommand) cmd)
+                    .filter(cmd -> cmd.getPlugin() == plugin).toList().forEach(cmd -> {
+                        cmd.unregister(cmdMap);
+                        cmds.values().removeIf(cmd::equals);
+                    })
+            )
+        );
+
+        // 5. Remove plugin from list
+        getPrivateField(pm, "plugins", new TypeToken<List<Plugin>>(){})
+            .ifPresent(plugins -> plugins.remove(plugin));
+
+        // 6. Remove lookup name
+        getPrivateField(pm, "lookupNames", new TypeToken<Map<String, Plugin>>(){})
+            .ifPresent(names -> names.remove(plugin.getName()));
+
+        // 7. Unload Java classes so that the jar can be deleted on Windows
+        if (plugin.getClass().getClassLoader() instanceof URLClassLoader cl)
+        {
+            if (!setPrivateField(cl, "plugin", null) || !setPrivateField(cl, "pluginInit", null))
+                System.err.println("Error in setPrivateField, skipping unload");
```

Comment:
:flushed: After all that work?

Reply:
This is just for unloading the classloader so that the jar file can be unlocked and deleted on Windows. Unix doesn't have this problem because Unix doesn't lock files.